diff -Naur binutils-2.37/bfd/config.bfd binutils-2.37-patched/bfd/config.bfd
--- binutils-2.37/bfd/config.bfd	2021-07-08 12:37:19.000000000 +0100
+++ binutils-2.37-patched/bfd/config.bfd	2021-12-01 07:12:32.602888671 +0000
@@ -634,7 +634,17 @@
     targ_selvecs=
     targ64_selvecs=x86_64_elf64_vec
     ;;
+  i[3-7]86-*-onyx*)
+    targ_defvec=i386_elf32_vec
+    targ_selvecs=
+    targ64_selvecs=x86_64_elf64_vec
+    ;;
 #ifdef BFD64
+  x86_64-*-onyx*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
   x86_64-*-cloudabi*)
     targ_defvec=x86_64_elf64_cloudabi_vec
     want64=true
diff -Naur binutils-2.37/config.sub binutils-2.37-patched/config.sub
--- binutils-2.37/config.sub	2021-07-08 12:37:19.000000000 +0100
+++ binutils-2.37-patched/config.sub	2021-12-01 07:12:32.602888671 +0000
@@ -1727,7 +1727,7 @@
 	     | scout* | superux* | sysv* | rtmk* | tpf* | windiss* \
 	     | powermax* | dnix* | nx6 | nx7 | sei* | dragonfly* \
 	     | skyos* | haiku* | rdos* | toppers* | drops* | es* \
-	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
+	     | onefs* | tirtos* | phoenix* | fuchsia* | onyx* | redox* | bme* \
 	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* \
 	     | nsk* | powerunix* | genode* | zvmoe* | qnx* | emx*)
 		;;
diff -Naur binutils-2.37/configure binutils-2.37-patched/configure
--- binutils-2.37/configure	2021-07-18 17:36:27.000000000 +0100
+++ binutils-2.37-patched/configure	2021-12-01 07:12:32.602888671 +0000
@@ -3013,7 +3013,7 @@
       *-*-elf* | *-*-sysv4* | *-*-unixware* | *-*-eabi* | hppa*64*-*-hpux* \
       | *-*-linux* | *-*-gnu* | frv-*-uclinux* | *-*-irix5* | *-*-irix6* \
       | *-*-netbsd* | *-*-openbsd* | *-*-freebsd* | *-*-dragonfly* \
-      | *-*-solaris2* | *-*-nto* | *-*-nacl*)
+      | *-*-solaris2* | *-*-nto* | *-*-nacl* | *-*-onyx*)
         case "${target}" in
           *-*-linux*aout* | *-*-linux*oldld*)
             ;;
diff -Naur binutils-2.37/configure.ac binutils-2.37-patched/configure.ac
--- binutils-2.37/configure.ac	2021-07-18 17:48:58.000000000 +0100
+++ binutils-2.37-patched/configure.ac	2021-12-01 07:12:32.606222005 +0000
@@ -336,7 +336,7 @@
       *-*-elf* | *-*-sysv4* | *-*-unixware* | *-*-eabi* | hppa*64*-*-hpux* \
       | *-*-linux* | *-*-gnu* | frv-*-uclinux* | *-*-irix5* | *-*-irix6* \
       | *-*-netbsd* | *-*-openbsd* | *-*-freebsd* | *-*-dragonfly* \
-      | *-*-solaris2* | *-*-nto* | *-*-nacl*)
+      | *-*-solaris2* | *-*-nto* | *-*-nacl* | *-*-onyx*)
         case "${target}" in
           *-*-linux*aout* | *-*-linux*oldld*)
             ;;
diff -Naur binutils-2.37/gas/configure.tgt binutils-2.37-patched/gas/configure.tgt
--- binutils-2.37/gas/configure.tgt	2021-07-08 12:37:19.000000000 +0100
+++ binutils-2.37-patched/gas/configure.tgt	2021-12-01 07:12:32.606222005 +0000
@@ -126,6 +126,7 @@
   aarch64*-*-elf*)			fmt=elf;;
   aarch64*-*-fuchsia*)			fmt=elf;;
   aarch64*-*-genode*)			fmt=elf;;
+  aarch64*-*-onyx*)			fmt=elf em=linux ;;
   aarch64*-*-linux*)			fmt=elf em=linux
     case ${cpu}-${os} in
       aarch64*-linux-gnu_ilp32)		arch=aarch64:32 ;;
@@ -222,6 +223,7 @@
   i386-*-elfiamcu)			fmt=elf arch=iamcu ;;
   i386-*-elf*)				fmt=elf ;;
   i386-*-fuchsia*)			fmt=elf ;;
+  i386-*-onyx*)         fmt=elf em=linux ;;
   i386-*-genode*)			fmt=elf ;;
   i386-*-bsd*)				fmt=aout em=386bsd ;;
   i386-*-netbsdpe*)			fmt=coff em=pe ;;
diff -Naur binutils-2.37/ld/configure.tgt binutils-2.37-patched/ld/configure.tgt
--- binutils-2.37/ld/configure.tgt	2021-07-08 12:37:20.000000000 +0100
+++ binutils-2.37-patched/ld/configure.tgt	2021-12-01 07:12:32.606222005 +0000
@@ -309,6 +309,10 @@
 i[3-7]86-*-rdos*)	targ_emul=elf_i386
 			targ_extra_emuls=elf_iamcu
 			;;
+x86_64-*-onyx*)
+			targ_emul=elf_x86_64_onyx
+			targ_extra_emuls="elf_i386_onyx elf_x86_64 elf_i386"
+			;;
 i[3-7]86-*-bsd)		targ_emul=i386bsd
 			targ_extra_ofiles=
 			;;
@@ -767,6 +771,10 @@
 			targ_extra_emuls="elf32briscv elf64lriscv elf32lriscv"
 			targ_extra_libpath=$targ_extra_emuls
 			;;
+riscv64*-*-onyx*)	targ_emul=elf64lriscv
+			targ_extra_emuls="elf64lriscv_lp64f elf64lriscv_lp64 elf32lriscv elf32lriscv_ilp32f elf32lriscv_ilp32"
+			targ_extra_libpath=$targ_extra_emuls
+			;;
 riscv64*-*-*)		targ_emul=elf64lriscv
 			targ_extra_emuls="elf32lriscv elf64briscv elf32briscv"
 			targ_extra_libpath=$targ_extra_emuls
diff -Naur binutils-2.37/ld/emulparams/elf_i386_onyx.sh binutils-2.37-patched/ld/emulparams/elf_i386_onyx.sh
--- binutils-2.37/ld/emulparams/elf_i386_onyx.sh	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.37-patched/ld/emulparams/elf_i386_onyx.sh	2021-12-01 07:12:32.606222005 +0000
@@ -0,0 +1,3 @@
+. ${srcdir}/emulparams/elf_i386.sh
+GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
\ No newline at end of file
diff -Naur binutils-2.37/ld/emulparams/elf_x86_64_onyx.sh binutils-2.37-patched/ld/emulparams/elf_x86_64_onyx.sh
--- binutils-2.37/ld/emulparams/elf_x86_64_onyx.sh	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.37-patched/ld/emulparams/elf_x86_64_onyx.sh	2021-12-01 07:12:32.606222005 +0000
@@ -0,0 +1 @@
+. ${srcdir}/emulparams/elf_x86_64.sh
\ No newline at end of file
diff -Naur binutils-2.37/ld/Makefile.am binutils-2.37-patched/ld/Makefile.am
--- binutils-2.37/ld/Makefile.am	2021-07-08 12:37:20.000000000 +0100
+++ binutils-2.37-patched/ld/Makefile.am	2021-12-01 07:12:32.606222005 +0000
@@ -282,6 +282,7 @@
 	eelf_i386_ldso.c \
 	eelf_i386_sol2.c \
 	eelf_i386_vxworks.c \
+	eelf_i386_onyx.c \
 	eelf_iamcu.c \
 	eelf_s390.c \
 	eh8300elf.c \
@@ -456,6 +457,7 @@
 	eelf_x86_64_cloudabi.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_sol2.c \
+	eelf_x86_64_onyx.c \
 	ehppa64linux.c \
 	ei386pep.c \
 	emmo.c
diff -Naur binutils-2.37/ld/Makefile.in binutils-2.37-patched/ld/Makefile.in
--- binutils-2.37/ld/Makefile.in	2021-07-18 17:38:34.000000000 +0100
+++ binutils-2.37-patched/ld/Makefile.in	2021-12-01 07:12:32.606222005 +0000
@@ -772,6 +772,7 @@
 	eelf_i386_ldso.c \
 	eelf_i386_sol2.c \
 	eelf_i386_vxworks.c \
+	eelf_i386_onyx.c \
 	eelf_iamcu.c \
 	eelf_s390.c \
 	eh8300elf.c \
@@ -945,6 +946,7 @@
 	eelf_x86_64_cloudabi.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_sol2.c \
+	eelf_x86_64_onyx.c \
 	ehppa64linux.c \
 	ei386pep.c \
 	emmo.c
@@ -1434,6 +1436,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_be.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_fbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_ldso.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_onyx.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_sol2.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_vxworks.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_iamcu.Po@am__quote@
@@ -1445,6 +1448,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_cloudabi.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_fbsd.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_onyx.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_sol2.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eh8300elf.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eh8300elf_linux.Po@am__quote@
diff -Naur binutils-2.37/PKGBUILD binutils-2.37-patched/PKGBUILD
--- binutils-2.37/PKGBUILD	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.37-patched/PKGBUILD	2021-12-01 07:12:32.606222005 +0000
@@ -0,0 +1,14 @@
+PKGNAME=binutils
+PKGVER=2.37
+
+build() {
+	cd ..
+	mkdir -p binutils-build
+	cd binutils-build
+	../binutils-2.37/configure --host=x86_64-onyx --prefix=/usr --with-sysroot=  --disable-werror --disable-nls --enable-gold --enable-lto --enable-plugins
+	$MAKE all
+}
+
+install() {
+	DESTDIR=$PKGDIR $MAKE install
+}
